pkg/ruler/manager.go

```diff
@@ -230,11 +233,11 @@ func (r *DefaultMultiTenantManager) getOrCreateNotifier(userID string) (*notifie
 
 func (r *DefaultMultiTenantManager) GetRules(userID string) []*promRules.Group {
 	var groups []*promRules.Group
-	r.userManagerMtx.Lock()
+	r.userManagerMtx.RLock()
 	if mngr, exists := r.userManagers[userID]; exists {
 		groups = mngr.RuleGroups()
```

Comment:
We don't need to protect `mngr.RuleGroups()` with the mutex, but just the lookup of `r.userManagers[userID]`. The manager already has an internal mutex, for example see the  `mngr.RuleGroups()` implementation:
https://github.com/prometheus/prometheus/blob/d56d0a9d52c8614fdf1c5c652061929ede562f3e/rules/manager.go#L1128-L1130

For this reason, I would just protect the access to `r.userManagers[userID]` with the mutex.